File: backend/fipm/routers/network.py
"""spec 11-nanopub-network.md §3.2: `GET /api/network/*`, the read-only
nanopublication-network proxy. All three routes are unauthenticated
(a workshop participant has no account) and read-only."""

from __future__ import annotations

import unicodedata
from datetime import UTC, datetime
from typing import Any

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from fipm import network
from fipm.config import Settings, get_settings
from fipm.db import get_db
from fipm.models import Fer

router = APIRouter(prefix="/network", tags=["network"])


def _require_network_enabled(settings: Settings) -> None:
    if not settings.network_enabled:
        raise HTTPException(status_code=503, detail="network_disabled")


def _to_http(exc: network.NetworkError) -> HTTPException:
    return HTTPException(status_code=exc.status_code, detail=exc.code)


def _iso(cached_at: float) -> str:
    return datetime.fromtimestamp(cached_at, tz=UTC).isoformat().replace("+00:00", "Z")


def _normalise(text: str) -> str:
    return unicodedata.normalize("NFC", text).casefold()


def _envelope(
    body: dict[str, Any], cached_at: float, stale: bool, settings: Settings
) -> dict[str, Any]:
    result = {**body, "cachedAt": _iso(cached_at), "source": settings.nanopub_query_url}
    if stale:
        result["stale"] = True
    return result


@router.get("/fip-communities")
def list_fip_communities(
    q: str | None = None,
    limit: int = 50,
    offset: int = 0,
) -> dict[str, Any]:
    settings = get_settings()
    _require_network_enabled(settings)
    limit = max(1, min(limit, 200))
    offset = max(0, offset)

    try:
        items, cached_at, stale = network.get_fip_communities(settings)
    except network.NetworkError as exc:
        raise _to_http(exc) from exc

    if q:
        needle = _normalise(q)
        items = [item for item in items if needle in _normalise(item["label"])]

    total = len(items)
    page = items[offset : offset + limit]
    return _envelope({"items": page, "total": total}, cached_at, stale, settings)


# ---------------------------------------------------------------------------
# GET /network/fips/{communityIri}: §3.3's shape, `inCatalogue` resolved here
# (against the `fers` table) since fipm.network is DB-free.
# ---------------------------------------------------------------------------


def _build_fer_indexes(db: Session) -> tuple[dict[str, Fer], dict[str, Fer], dict[str, Fer]]:
    """spec §3.3: the three lookup indexes `_match_in_catalogue` tries, in
    priority order -- by `Fer.id`, by homepage (`Fer.homepage` or
    `Fer.id` itself, the `skos:exactMatch` case), by exact case-folded
    label (one entry per `|`-separated part of `Fer.label_search`)."""
    by_id: dict[str, Fer] = {}
    by_homepage: dict[str, Fer] = {}
    by_label: dict[str, Fer] = {}
    for row in db.query(Fer).all():
        by_id[row.id] = row
        if row.homepage:
            by_homepage.setdefault(row.homepage, row)
        by_homepage.setdefault(row.id, row)
        for part in (row.label_search or "").split("|"):
            if part:
                by_label.setdefault(part, row)
    return by_id, by_homepage, by_label


def _match_in_catalogue(
    resource: dict[str, Any],
    by_id: dict[str, Fer],
    by_homepage: dict[str, Fer],
    by_label: dict[str, Fer],
) -> dict[str, str] | None:
    row = by_id.get(resource["iri"])
    if row is not None:
        return {"ferId": row.id, "matchedBy": "ferId"}
    homepage = resource.get("homepage")
    if homepage:
        row = by_homepage.get(homepage)
        if row is not None:
            return {"ferId": row.id, "matchedBy": "homepage"}
    label = resource.get("label")
    if label:
        row = by_label.get(label.casefold())
        if row is not None:
            return {"ferId": row.id, "matchedBy": "label"}
    return None


def _resolve_in_catalogue(db: Session, payload: dict[str, Any]) -> None:
    by_id, by_homepage, by_label = _build_fer_indexes(db)
    for bucket in (*payload["questions"], *payload["unmapped"]):
        for decl in bucket["declarations"]:
            resource = decl.get("resource")
            if resource is not None:
                resource["inCatalogue"] = _match_in_catalogue(
                    resource, by_id, by_homepage, by_label
                )


@router.get("/fips/{community_iri:path}")
def get_network_fip(community_iri: str, db: Session = Depends(get_db)) -> dict[str, Any]:
    settings = get_settings()
    _require_network_enabled(settings)

    try:
        payload, cached_at, stale = network.get_community_fip(settings, community_iri)
    except network.NetworkError as exc:
        raise _to_http(exc) from exc

    # The community's own label isn't part of Q1/Q2/Q3 -- it's the Q0 list's
    # job (spec §3.1). Best-effort lookup against the (cached) community
    # list; None if the upstream call fails or the community isn't in it
    # (shouldn't happen for a community that just answered Q1, but a 502
    # here must never turn a working fip-detail response into a failure).
    label = None
    try:
        communities, _, _ = network.get_fip_communities(settings)
    except network.NetworkError:
        communities = []
    for item in communities:
        if item["iri"] == community_iri:
            label = item["label"]
            break
    payload["community"]["label"] = label

    _resolve_in_catalogue(db, payload)

    return _envelope(payload, cached_at, stale, settings)


@router.get("/fers")
def search_network_fers(q: str = "", limit: int = 20) -> dict[str, Any]:
    """v2 roadmap (spec §7 Q8): shipped and testable, not yet wired into
    `FerPicker.vue`."""
    settings = get_settings()
    _require_network_enabled(settings)
    limit = max(1, min(limit, 200))

    try:
        items, cached_at, stale = network.search_fers(settings, q, limit)
    except network.NetworkError as exc:
        raise _to_http(exc) from exc

    return _envelope({"items": items, "total": len(items)}, cached_at, stale, settings)

File: frontend/src/types/network.ts
/**
 * DTOs for the nanopublication-network read endpoints (spec 11 §3.2/§3.3)
 * and the local nanopublication-export bundle manifest (spec 11 §2.6).
 * Wire shape is camelCase throughout, same convention as `types/api.ts`.
 */

import type { DeclarationStatus, Visibility } from './api'

// ---------------------------------------------------------------------------
// spec 11 §3.2: GET /api/network/fip-communities
// ---------------------------------------------------------------------------

export interface NetworkCommunitySummary {
  iri: string
  label: string
  fipCount: number
}

export interface NetworkCommunityListResponse {
  items: NetworkCommunitySummary[]
  total: number
  cachedAt: string
  /** `settings.nanopub_query_url`, added by every `_envelope`-wrapped response (`routers/network.py`). */
  source: string
  /** Present on a stale cache hit served after an upstream failure (spec 11 §3.4). */
  stale?: boolean
}

// ---------------------------------------------------------------------------
// spec 11 §3.2: GET /api/network/fers (v2 roadmap item, shipped behind the
// same flag but not wired into the FER picker for CONFOA — spec 11 §3.2).
// ---------------------------------------------------------------------------

export interface NetworkFerSummary {
  iri: string
  /** `null` when Q4's own hit label and the Q3 enrichment both came back empty (`network.py`'s `search_fers`). */
  label: string | null
  comment: string | null
  homepage: string | null
  types: string[]
  ferTypeKey: string | null
}

export interface NetworkFerSearchResponse {
  items: NetworkFerSummary[]
  total: number
}

// ---------------------------------------------------------------------------
// spec 11 §3.3: GET /api/network/fips/{communityIri}
// ---------------------------------------------------------------------------

/** spec 11 §3.3: filled by matching the network resource against our `fers` table. */
export interface NetworkInCatalogue {
  ferId: string
  matchedBy: string
}

export interface NetworkResource {
  iri: string
  /** `null` when Q3 never returned a row for this resource IRI (`network.py`'s `_group_resources`/`get_community_fip`). */
  label: string | null
  comment: string | null
  homepage: string | null
  types: string[]
  ferTypeKey: string | null
  inCatalogue: NetworkInCatalogue | null
}

export interface NetworkDeclaration {
  nanopubIri: string
  /**
   * Inverse of `rdf.STATUS_PREDICATE`, plus `?nochoice → "none"` — same
   * vocabulary as `DeclarationStatus`. `null` when the declaration is
   * neither a no-choice declaration nor carries any of the four
   * `declares-*-use-of`/`declares-*-replacement-of`/`declares-*-development-of`
   * predicates (`network.py`'s `_status_and_resource`) — render the same
   * fallback used for `resource === null` rather than passing `null` into
   * `StatusBadge`, which has no entry for it.
   */
  status: DeclarationStatus | null
  /**
   * `null` for a "no choice" declaration (spec 11 §3.3's `?nochoice` case):
   * `{nanopubIri, status: "none", resource: null, considerations: null,
   * startDate: null, endDate: null}`. Render as `network.noChoiceDeclared`
   * rather than dereferencing `.label`.
   */
  resource: NetworkResource | null
  considerations: string | null
  startDate: string | null
  endDate: string | null
}

export interface NetworkQuestion {
  questionId: string
  questionIri: string
  declarations: NetworkDeclaration[]
}

export interface NetworkUnmappedQuestion {
  questionIri: string
  declarations: NetworkDeclaration[]
}

export interface NetworkFipOtherVersion {
  nanopubIri: string
  /** `null` when the upstream row's `?label` binding is missing (`network._val`). */
  label: string | null
  /** `null` when the upstream row's `?created` binding is missing (`network._val`). */
  created: string | null
}

export interface NetworkFipSummary {
  /** `null` when the upstream row's `?fip_np` binding is missing (`network._val`). */
  nanopubIri: string | null
  /** `null` when the upstream row's `?label` binding is missing (`network._val`). */
  label: string | null
  /** `null` when the FIP nanopub carries no `fip:has-declaration-index` (optional in Q1). */
  indexIri: string | null
  /** `null` when the upstream row's `?created` binding is missing (`network._val`). */
  created: string | null
  startDate: string | null
  endDate: string | null
  otherVersions: NetworkFipOtherVersion[]
}

export interface NetworkCommunityFipResponse {
  /** `label` is a best-effort lookup against the community list (`routers/network.py`'s `get_network_fip`) — `null` on any upstream failure or an IRI no longer in that list. */
  community: { iri: string; label: string | null }
  fip: NetworkFipSummary
  /** All 21, in `gofair-fip-mini` order, each with a possibly empty `declarations` (spec 11 §3.3). */
  questions: NetworkQuestion[]
  /** Anything not among the 21 (an `FIP-S-Question-*`, an `FSR` question, a future 22nd question). */
  unmapped: NetworkUnmappedQuestion[]
  cachedAt: string
  source: string
  stale?: boolean
}

// ---------------------------------------------------------------------------
// spec 11 §3.6: POST /api/fips/from-network
// ---------------------------------------------------------------------------

export interface CreateFipFromNetworkRequest {
  communityIri: string
  title?: string
  language?: string
  visibility?: Visibility
  /**
   * A facilitator-session deep link (`?session=<id>&code=<joinCode>` on
   * `NetworkFipDetail.vue`, spec 11 §3.6) — both present together or both
   * omitted, never one without the other.
   */
  sessionId?: string
  joinCode?: string
}

/**
 * `fip` is the normal `FipOut` payload; kept as `Record<string, unknown>`
 * here (rather than importing `FipOut`) only to avoid a circular-looking
 * import for a shape callers re-parse anyway via `getFip`/`FipOut` once
 * navigated — callers that need the typed FIP use `FipOut` from `types/api`.
 */
export interface CreateFipFromNetworkResponse {
  fip: { id: string } & Record<string, unknown>
  /** Present exactly once, for an anonymous/standalone creation (same rule as `POST /api/fips`). */
  editToken?: string
  imported: { declarations: number; fersCreated: number; fersMatched: number }
  skipped: Array<{ questionIri: string; reason: string }>
}

// ---------------------------------------------------------------------------
// spec 11 §2.6: nanopublication export bundle — index.json
// ---------------------------------------------------------------------------

export interface NanopubBundleGenerator {
  tool: string
  agent: string
  generatedAt: string
}

export interface NanopubBundleFip {
  id: string
  url: string
  title: string
  language: string
  license: string
  questionnaire: { id: string; version: string }
}

export interface NanopubEntry {
  n: number
  role: string
  file: string
  tempNanopubIri: string
  /** Explicit `null` (not omitted) for the index entry, which has no concept of its own (`nanopub_export.py`). */
  conceptIri: string | null
  introduces?: string
  referencedBy?: number[]
  questionId?: string
  questionIri?: string
  declarationIndex?: number
  status?: string
  /** Explicit `null` (not omitted) for a declaration with no resource (a no-choice declaration). */
  resourceIri: string | null
  /** Explicit `null` (not omitted) alongside `resourceIri: null`. */
  resourceOrigin: string | null
  references?: number[]
}

export interface NanopubRewrite {
  afterSigning: number
  replacePrefix: string
  withPrefix: string
  inFiles: string[]
}

export interface NanopubCounts {
  nanopubs: number
  declarations: number
  skipped: number
}

export interface NanopubSkipped {
  questionId: string
  declarationIndex: number
  reason: string
}

export interface NanopubDmpEvidenceGap {
  questionId: string
  declarationIndex: number
}

export interface NanopubNotRepresented {
  notApplicable: string[]
  answerComments: string[]
  dmpEvidence: NanopubDmpEvidenceGap[]
  relatedDmps: number
}

export interface NanopubPrerequisite {
  what: string
  detail: string
}

export interface NanopubPublishPrerequisites {
  satisfied: NanopubPrerequisite[]
  missing: NanopubPrerequisite[]
}

export interface NanopubBundleIndex {
  schema: string
  generator: NanopubBundleGenerator
  fip: NanopubBundleFip
  tempBase: string
  signed: boolean
  nanopubs: NanopubEntry[]
  signingOrder: Array<number | string>
  rewrites: NanopubRewrite[]
  counts: NanopubCounts
  skipped: NanopubSkipped[]
  notRepresented: NanopubNotRepresented
  publishPrerequisites: NanopubPublishPrerequisites
}
